### PR TITLE
Select text using douglas, rainbow, and legion

### PR DESCRIPTION
**Pull request by seekM**
This PR allows for text selection using douglas and rainbow. There are three ways to select text (using douglas as an example here, rainbow is analogous):
- long version: `<x1> [by] <y1> select <x2> [by] <y2>`
- short version: `<x1> [by] <y1> select <x2>` (with an implicit `y2 = y1` value). This is useful to select text which is on one horizontal line
- point version: set two points first (`point one`, `point two`), then do a `select`. This allows for precise text selection. How to do it in detail? For the first point you can do `<x1> [by] <y1> move`, the fine-tune via `curse`. Then do `point one`. Then do the same for the second point, then `select`.

Furthermore I replaced `Playback([(["mouse", "right", "click"], 0.0)]).execute()` by `Mouse("right").execute()` (same for left click). This makes the code more language agnostic. Dragonfly offers a mouse action to drag the mouse, but it didn't work for me. Therefore I used `win32api, win32con` directly. Does `grids.wait_for_death(settings.DOUGLAS_TITLE)` work? I couldn't track down the respective setting / value and had to add `time.sleep(0.1)` at one point as a workaround.

### CODE REVIEW

File: castervoice/apps/griddouglas.py [hunk 1]
```diff
@@ -63,4 +121,4 @@ class GridControlRule(MergeRule):
     rule = GridControlRule(name="Douglas")
     gfilter.run_on(rule)
     grammar.add_rule(rule)
-    grammar.load()
+    grammar.load()
```
Comment (kendonB) [inline castervoice/apps/griddouglas.py]: Does no newline matter for python files?

File: castervoice/apps/griddouglas.py [hunk 2]
```diff
@@ -13,47 +13,105 @@
...
+        IntegerRefST("x", 0, 300),
+        IntegerRefST("y", 0, 300),
+        IntegerRefST("x1", 0, 300),
+        IntegerRefST("y1", 0, 300),
+        IntegerRefST("x2", 0, 300),
+        IntegerRefST("y2", 0, 300),
         Choice("action", {
             "kick": 0,
             "psychic": 1,
+            "move": 2,
```
Comment (kendonB) [inline castervoice/apps/griddouglas.py]: "hover"? Does caster have a standard spec for other hover/move commands?

**Comment by LexiconCode** [comment]
@seekM could you update the documentation and provide a few example commands here. The grid seems to work in general but it's hard to formulate testing looking straight from the code.

**Comment by LexiconCode** [comment]
@kendonB could you test this as well?

File: castervoice/apps/griddouglas.py (hunk 1 above)
Comment (lexxish) [reply castervoice/apps/griddouglas.py]: Won't break anything but it should be added

File: castervoice/apps/griddouglas.py (hunk 2 above)
Comment (lexxish) [reply castervoice/apps/griddouglas.py]: "light" https://github.com/dictation-toolbox/Caster/blob/c5074e2a8b7de348cbfa80a4686e7e2d5db2ed9a/caster/apps/gridlegion.py#L73

**Comment by seekM** [comment]
Unfortunately I'm busy with other things at the moment, I'll add doumentation later.

**Comment by kendonB** [comment]
I've tried this and it works beautifully. Note that the order of calling when using the Douglas grid is Y then X.

The process for the fine selection option should be explained a bit more clearly too.

Fantastic contribution, I am going to merge this one into my own copy right away.

File: castervoice/apps/griddouglas.py [hunk 3]
```diff
@@ -13,47 +13,105 @@
...
+        }),
+        Choice("point", {
+            "one": 1,
+            "two": 2,
         }),
     ]
     defaults = {
-        "action": -1,
+        "action": 0,
```
Comment (kendonB) [inline castervoice/apps/griddouglas.py]: Why does the default change?

File: castervoice/apps/griddouglas.py (hunk 2 above)
Comment (kendonB) [reply castervoice/apps/griddouglas.py]: light in legion highlights the word.

**Comment by kendonB** [comment]
@seekM can you test my addition for legion?

File: castervoice/apps/griddouglas.py (hunk 3 above)
Comment (kendonB) [reply castervoice/apps/griddouglas.py]: This changes the default to click when it should be move (-1)

**Comment by kendonB** [comment]
@seekM were you able to review my changes? please test the functionality of the Legion selection and take a look at my changes to the documentation. We will be able to merge after that.

**Comment by seekM** [comment]
Legion doesn't work for me at the moment, so I can't test it. Also when I tried to checkout this PR there were some errors which are due to the files in my Caster user folder I think. I don't have time for this at the moment so it would be cool if someone else could review if necessary.

**Comment by kendonB** [comment]
> Also when I tried to checkout this PR there were some errors which are due to the files in my Caster user folder I think.

Could you paste the errors here?

**Comment by seekM** [comment]
```
Ignoring ccr rule 'Navigation'. Failed to load with: 
<type 'exceptions.TypeError'>
executable argument must be a string or None; received ['rstudio', 'foxitreader']
<traceback object at 0x0F410288>
  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\__init__.py", line 35, in <module>
    [class_name])  # attempts to import the class

  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\core\nav.py", line 165, in <module>
    class Navigation(MergeRule):

  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\core\nav.py", line 229, in Navigation
    (AppContext(executable=["rstudio", "foxitreader"]), Key("cs-z")*Repeat(extra="nnavi10")),

  File "C:\dev\Python27\lib\site-packages\dragonfly\grammar\context.py", line 217, in __init__
    " received %r" % executable)

Ignoring ccr rule 'HTML'. Failed to load with: 
<type 'exceptions.TypeError'>
executable argument must be a string or None; received ['idea', 'idea64', 'studio64', 'pycharm', 'webstorm64', 'webstorm']
<traceback object at 0x0F9FB3A0>
  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\__init__.py", line 35, in <module>
    [class_name])  # attempts to import the class

  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\html\html.py", line 11, in <module>
    context = AppContext(executable=["idea", "idea64", "studio64", "pycharm", "webstorm64", "webstorm"])

  File "C:\dev\Python27\lib\site-packages\dragonfly\grammar\context.py", line 217, in __init__
    " received %r" % executable)

Ignoring application 'eclipse'. Failed to load with: 
<type 'exceptions.TypeError'>
executable argument must be a string or None; received ['rstudio', 'foxitreader']
<traceback object at 0x108A8A08>
  File "C:\dev\Natlink-User\Caster\castervoice\apps\__init__.py", line 13, in is_valid
    _ = __import__(module, globals(), locals())

  File "C:\dev\Natlink-User\Caster\castervoice\apps\eclipse.py", line 15, in <module>
    from castervoice.lib.ccr.core.nav import Navigation

  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\core\nav.py", line 165, in <module>
    class Navigation(MergeRule):

  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\core\nav.py", line 229, in Navigation
    (AppContext(executable=["rstudio", "foxitreader"]), Key("cs-z")*Repeat(extra="nnavi10")),

  File "C:\dev\Python27\lib\site-packages\dragonfly\grammar\context.py", line 217, in __init__
    " received %r" % executable)

Ignoring application 'flashdevelop'. Failed to load with: 
<type 'exceptions.TypeError'>
executable argument must be a string or None; received ['rstudio', 'foxitreader']
<traceback object at 0x10AC9468>
  File "C:\dev\Natlink-User\Caster\castervoice\apps\__init__.py", line 13, in is_valid
    _ = __import__(module, globals(), locals())

  File "C:\dev\Natlink-User\Caster\castervoice\apps\flashdevelop.py", line 6, in <module>
    from castervoice.lib.ccr.core.nav import Navigation

  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\core\nav.py", line 165, in <module>
    class Navigation(MergeRule):

  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\core\nav.py", line 229, in Navigation
    (AppContext(executable=["rstudio", "foxitreader"]), Key("cs-z")*Repeat(extra="nnavi10")),

  File "C:\dev\Python27\lib\site-packages\dragonfly\grammar\context.py", line 217, in __init__
    " received %r" % executable)

Error loading _caster from C:\dev\Natlink-User\Caster\_caster.py
Traceback (most recent call last):
  File "C:\dev\NatLink\NatLink\MacroSystem\core\natlinkmain.py", line 331, in loadFile
    imp.load_module(modName,fndFile,fndName,fndDesc)
  File "C:\dev\Natlink-User\Caster\_caster.py", line 44, in <module>
    import castervoice.lib.dev.dev
  File "C:\dev\Natlink-User\Caster\castervoice\lib\dev\__init__.py", line 1, in <module>
    import dev
  File "C:\dev\Natlink-User\Caster\castervoice\lib\dev\dev.py", line 21, in <module>
    from castervoice.lib.tests import testrunner
  File "C:\dev\Natlink-User\Caster\castervoice\lib\tests\testrunner.py", line 3, in <module>
    from castervoice.lib.tests.unit import merger, mergerule, filter, state, stack, node, textformat
  File "C:\dev\Natlink-User\Caster\castervoice\lib\tests\unit\merger.py", line 3, in <module>
    from castervoice.apps import eclipse
  File "C:\dev\Natlink-User\Caster\castervoice\apps\eclipse.py", line 15, in <module>
    from castervoice.lib.ccr.core.nav import Navigation
  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\core\nav.py", line 165, in <module>
    class Navigation(MergeRule):
  File "C:\dev\Natlink-User\Caster\castervoice\lib\ccr\core\nav.py", line 229, in Navigation
    (AppContext(executable=["rstudio", "foxitreader"]), Key("cs-z")*Repeat(extra="nnavi10")),
  File "C:\dev\Python27\lib\site-packages\dragonfly\grammar\context.py", line 217, in __init__
    " received %r" % executable)
TypeError: executable argument must be a string or None; received ['rstudio', 'foxitreader']
-- skip unchanged wrong grammar file: C:\dev\Natlink-User\Caster\_caster.py
--- userLanguage: deu

-- skip unchanged wrong grammar file: C:\dev\Natlink-User\Caster\_caster.py
natlinkmain started from C:\dev\NatLink\NatLink\MacroSystem\core:
```

**Comment by kendonB** [comment]
Try updating dragonfly2

**Comment by seekM** [comment]
Updating dragonfly2 made these errors go away

**Comment by LexiconCode** [comment]
ready to merge?

**Comment by kendonB** [comment]
legion behavior needs testing by someone other than me for whom legion is working.

**Comment by kendonB** [comment]
I'd also like someone to review the documentation changes.

File: castervoice/doc/readthedocs/Mouse.md [hunk 4]
```diff
@@ -38,34 +38,48 @@ Demonstration [here](https://youtu.be/UISjQBMmQ-I).
...
-**Example**:
+**Examples**:
 
-- _red 86_: - References the according square within the first color palette.
-- _two red 86_ - References the according square within the second color palette.
+- _red 86_: - References square within the first red palette.
+- _red 86 kick_: - Clicks at this location.
+- _two red 86_ - References square within the second red palette.
+- _red 86 select 2 green 10_: - Selects from the first red 86 to the second green 10.
+- _red 86, curse sauce lease 5, point one, select 2 green 10, curse dunce lease 10, point two, select_: - Selects from the first red 86 to the second green 10 with fine adjustment.
```
Comment (codingApprentice) [inline castervoice/doc/readthedocs/Mouse.md]: Should it be, 
"_red 86 move, curse sauce lease 5, point one, 2 green 10 move, curse dunce lease 10, point two, select_"

**Comment by shervinemami** [comment]
Nice work guys! I implemented a similar feature in recent weeks using Aenea on Linux based on Caster's Douglas grid, I haven't submitted it to the Caster project because my method currently only supports Linux + Aenea, but I've been using it about 50 times per day and it's been very useful! I just thought I'd mention that instead of your "point 1" & "point 2" keywords, you might like to use the keywords "drag" & "release", if you find it more intuitive? In my BeamGrid system I move the mouse to a starting point, I say "drag", then move to the end point, then say "release".

**Comment by kendonB** [comment]
@codingApprentice, good catch and thanks for having a look!

@shervinemami, great point; note that the caster commands "squat" and "bench" are used for holding and releasing the left mouse button.

Since Caster now supports Aenea (or is trying to) it would be great for you to to add your code here! The mouse grids are the only thing preventing me from trying out Caster with Aenea at this point.

**Comment by kendonB** [comment]
I don't see any value in using "point two, select" instead of using "bench" for both. Can anyone think of a reason to keep them separate?

**Comment by kendonB** [comment]
The behavior between Douglas grid and rainbow grid for the default action is also different. I think it should work like rainbow does now which automatically moves the cursor and closes the grid.

**Comment by LexiconCode** [comment]
ready to merge?

**Comment by shervinemami** [comment]
> Since Caster now supports Aenea (or is trying to) it would be great for you to to add your code here! The mouse grids / legion are the only thing preventing me from trying out Caster with Aenea at this point.

Oh I didn't know Caster is trying to support Aenea. I'll discuss that on the gitter forum.

File: castervoice/doc/readthedocs/Mouse.md [hunk 5]
```diff
@@ -26,46 +26,60 @@ Demonstration [here](https://youtu.be/UISjQBMmQ-I).
...
 **Usage**:
 
-- Evoke: `rainbow`
-- Select square: `[<number_of_color_palette] <color> <number>`  
+- Invoke: `rainbow` or `rainbow <monitor_number>` 
+- Move to square: `[<number_of_color_palette>] <color> <number>`  
   _number_of_color_palette_ refers to the fact that colors loop when filling the screen.  
   _color_ can be replaced by the following colors: _red_, _orange_, _yellow_, _green_, _blue_, _purple_.
+- Select: `[<number_of_color_palette_1>] <color_1> <number_1>` select `[<number_of_color_palette_2>] <color_2> <number_2>`
+- Fine selection: `[<number_of_color_palette_1>] <color_1> <number_1> move` &rightarrow; `curse ...` &rightarrow; `point one` &rightarrow; `[<number_of_color_palette_2>] <color_2> <number_2> move` &rightarrow; `curse ...` &rightarrow; `point two` &rightarrow; `select`
```
Comment (kendonB) [inline castervoice/doc/readthedocs/Mouse.md]: update docs

File: castervoice/doc/readthedocs/Mouse.md [hunk 6]
```diff
@@ -26,46 +26,60 @@ Demonstration [here](https://youtu.be/UISjQBMmQ-I).
...
 
 **Usage**:
 
-- Evoke: `douglas`
-- Select square: `<horizontal_number> by <vertical_number>`
+- Invoke: `douglas` or `douglas <monitor_number>`
+- Move to square: `<horizontal_number> by <vertical_number>`
+- Select: `<horizontal_number_1> by <vertical_number_1> select <horizontal_number_2> by <vertical_number_2>`
+- Select horizontally: `<horizontal_number_1> by <vertical_number_1> select <horizontal_number_2>`
+- Fine selection: `<horizontal_number_1> by <vertical_number_1> move` &rightarrow; `curse ...` &rightarrow; `point one` &rightarrow; `<horizontal_number_2> by <vertical_number_2> move` &rightarrow; `curse ...` &rightarrow; `point two` &rightarrow; `select`
```
Comment (kendonB) [inline castervoice/doc/readthedocs/Mouse.md]: need to change docs here

File: castervoice/doc/readthedocs/Mouse.md [hunk 7]
```diff
@@ -75,5 +89,12 @@ Demonstration [here](https://youtu.be/UISjQBMmQ-I).
 
 **Usage**:
 
-- Evoke: `legion`
-- Select text area: `<number>`
\ No newline at end of file
+- Invoke: `legion or legion [monitor_number]`
+- Move to text area: `<number>`
+- Highlight one text area: `<number> (select | light)`
+- Highlight from one text area to another: `<number_1> (select | light) <number_2>`
```
Comment (kendonB) [inline castervoice/doc/readthedocs/Mouse.md]: add grab here

**Pull request merged**